common/models/templates/src/trie-model.ts

```diff
@@ -161,6 +161,116 @@
     public wordbreak(context: Context): USVString {
       return this.getLastWord(context.left);
     }
+
+    public getRootTraversal(): LexiconTraversal {
+      return new TrieModel.Traversal(this._trie['root'], '');
+    }
+
+    private static Traversal = class implements LexiconTraversal {
+      prefix: String;
+      root: Node;
+
+      constructor(root: Node, prefix: string) {
+        this.root = root;
+        this.prefix = prefix;
+      }
+
+      *children(): Generator<{char: string, traversal: () => LexiconTraversal}> {
+        let root = this.root;
+
+        if(root.type == 'internal') {
+          for(let i = 0; i < root.values.length; i++) {
+            let entry = root.values[i];
+            let entryNode = root.children[entry];
+
+            // SMP check
+            let charCode = entry.charCodeAt(0);
+            if(charCode >= 0xD800 && charCode <= 0xDBFF) {
```

Comment:
Maybe factor out into a function?

```suggestion
            if(isHighSurrogate(charCode)) {
```